TableLink is broken for Oracle database after pull request #2903

User name and password are not passed to the RDBMS after request #2903 

Below is a test case to reproduce the problem for Oracle and the corresponding stack dump. I assume that it would be the same for all other databases (Teradata/DB2 and so on). I noticed that there are no tests in TestLinkedTable against databases other than H2. I suggest to all such tests and make them optional because they require additional external databases.

```
    private void testOracle() throws SQLException, ClassNotFoundException {
        deleteDb("linkedTableOracle");
        String driver = "oracle.jdbc.driver.OracleDriver";
        Class.forName(driver);
        String url = "jdbc:oracle:thin:@//server:1521/ORCLPDB1.localdomain";
        Connection oracle = DriverManager.getConnection(url,
                "user", "pass");
        oracle.setAutoCommit(false);

        Connection connMain = DriverManager.getConnection("jdbc:h2:mem:linkedTableOracle");

        Statement ora = oracle.createStatement();
        if (false) {
            // table was created before the test
            ora.execute("CREATE TABLE ORA (DT DATE)");
            ora.execute("INSERT INTO ORA VALUES (TIMESTAMP'2008-01-08 10:11:15')");
        }
        oracle.commit();

        Statement statMain = connMain.createStatement();

        statMain.execute(MessageFormat.format("CREATE LINKED TABLE LINK(''{0}'', ''{1}'', ''user'', ''pass'', ''ORA'');",
                "oracle.jdbc.driver.OracleDriver", url));

        ResultSet rs = statMain.executeQuery("select * FROM LINK");
        rs.next();
        assertEquals(rs.getDate(1).toString(), "2008-01-08");
        assertFalse(rs.next());

        connMain.close();
    }
```


```
Exception in thread "main" org.h2.jdbc.JdbcSQLException: ORA-01017: invalid username/password; logon denied
; SQL statement:
CREATE LINKED TABLE LINK('oracle.jdbc.driver.OracleDriver', 'jdbc:oracle:thin:@//server:1521/ORCLPDB1.localdomain', 'user', 'pass', 'ORA'); [1017-201]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:661)
	at org.h2.message.DbException.addSQL(DbException.java:154)
	at org.h2.command.Command.executeUpdate(Command.java:266)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:235)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:206)
	at org.h2.test.db.TestLinkedTable.testOracle(TestLinkedTable.java:84)
	at org.h2.test.db.TestLinkedTable.test(TestLinkedTable.java:40)
	at org.h2.test.TestBase.testFromMain(TestBase.java:472)
	at org.h2.test.db.TestLinkedTable.main(TestLinkedTable.java:35)
Caused by: java.sql.SQLException: ORA-01017: invalid username/password; logon denied

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:494)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:441)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:436)
	at oracle.jdbc.driver.T4CTTIfun.processError(T4CTTIfun.java:1061)
	at oracle.jdbc.driver.T4CTTIoauthenticate.processError(T4CTTIoauthenticate.java:550)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:623)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:252)
	at oracle.jdbc.driver.T4CTTIoauthenticate.doOAUTH(T4CTTIoauthenticate.java:499)
	at oracle.jdbc.driver.T4CTTIoauthenticate.doOAUTH(T4CTTIoauthenticate.java:1279)
	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:663)
	at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:688)
	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:39)
	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:691)
	at org.h2.util.JdbcUtils.getConnection(JdbcUtils.java:299)
	at org.h2.util.JdbcUtils.getConnection(JdbcUtils.java:265)
	at org.h2.table.TableLinkConnection.open(TableLinkConnection.java:89)
	at org.h2.table.TableLinkConnection.open(TableLinkConnection.java:76)
	at org.h2.engine.Database.getLinkConnection(Database.java:2508)
	at org.h2.table.TableLink.connect(TableLink.java:98)
	at org.h2.table.TableLink.<init>(TableLink.java:81)
	at org.h2.schema.Schema.createTableLink(Schema.java:830)
	at org.h2.command.ddl.CreateLinkedTable.update(CreateLinkedTable.java:77)
	at org.h2.command.CommandContainer.update(CommandContainer.java:173)
	at org.h2.command.Command.executeUpdate(Command.java:250)

```

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #2958: TableLink is broken for Oracle database after pull request #2903
+</li>
 <li>PR #2955: Prevent incorrect index sorting
 </li>
 <li>PR #2951: Add documentation for INFORMATION_SCHEMA

--- a/h2/src/main/org/h2/util/JdbcUtils.java
+++ b/h2/src/main/org/h2/util/JdbcUtils.java
@@ -27,6 +27,8 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Comparator;
 import java.util.HashSet;
+import java.util.Properties;
+
 import javax.naming.Context;
 import javax.sql.DataSource;
 import org.h2.api.ErrorCode;
@@ -291,12 +293,19 @@ public class JdbcUtils {
             try {
                 if (java.sql.Driver.class.isAssignableFrom(d)) {
                     Driver driverInstance = (Driver) d.getDeclaredConstructor().newInstance();
+                    Properties prop = new Properties();
+                    if (user != null) {
+                        prop.setProperty("user", user);
+                    }
+                    if (password != null) {
+                        prop.setProperty("password", password);
+                    }
                     /*
                      * fix issue #695 with drivers with the same jdbc
                      * subprotocol in classpath of jdbc drivers (as example
                      * redshift and postgresql drivers)
                      */
-                    Connection connection = driverInstance.connect(url, null);
+                    Connection connection = driverInstance.connect(url, prop);
                     if (connection != null) {
                         return connection;
                     }

